clang/lib/3C/RewriteUtils.cpp

```diff
@@ -128,10 +128,56 @@ GlobalVariableGroups::~GlobalVariableGroups() {
 // Note that R.getRangeSize will return -1 if SR is within
 // a macro as well. This means that we can't re-write any
 // text that occurs within a macro.
-bool canRewrite(Rewriter &R, SourceRange &SR) {
+bool canRewrite(Rewriter &R, const CharSourceRange &SR) {
   return SR.isValid() && (R.getRangeSize(SR) != -1);
 }
 
+void rewriteSourceRange(Rewriter &R, const SourceRange &Range,
+                        const std::string &NewText) {
+  rewriteSourceRange(R, CharSourceRange::getTokenRange(Range), NewText);
+}
+
+void rewriteSourceRange(Rewriter &R, const CharSourceRange &Range,
+                        const std::string &NewText) {
+  if (canRewrite(R, Range)) {
+    R.ReplaceText(Range, NewText);
+  } else {
+    // There is a macro somewhere in the source range, but it wasn't noticed
+    // during constraint generation. This typically means that the only part of
+    // the range being rewritten is in a macro. For declarations, this seems to
+    // mean that the identifier is not in a macro, but other parts of the
+    // declaration might be.
+    CharSourceRange Expand = R.getSourceMgr().getExpansionRange(Range);
```

Comment:
My tests suggest that this code replaces the entire macro call even if the macro expansion included tokens that are not part of what we want to replace.  Some examples:
```c
// Same as the original example of #429 except the semicolon is inside the macro
#define args ();
typedef int (*a) args
a b;

#define MIDDLE x; int *
int MIDDLE y;
```
The code currently in this PR issues the "Rewriting might clobber macro" warning and then gives the following incorrect output:
```c
#define args ();
typedef _Ptr<int (void)> a    // lost the semicolon
a b = ((void *)0);

#define MIDDLE x; int *
int _Ptr<int> y = ((void *)0);    // lost the "x; "
```

If you instead do this (similar to what CastPlacementVisitor is doing):
```suggestion
    CharSourceRange Expand = clang::Lexer::makeFileCharRange(
        Range, R.getSourceMgr(), R.getLangOpts());
```
then my bad examples produce the "Unable to rewrite converted source range" error, while the example of #429 works the same way as in the current PR.

My examples are contrived, yes, but it doesn't seem to cost us much to get this right.  (Or closer to right?  This stuff is bending my mind.)

Reply:
@john-h-kastner Perhaps try @mattmccutchen-cci 's fix and confirm nothing else breaks?

(I agree that this is hard to think about, and I've barely dipped my toe in!)

Reply:
Woohoo!  John, will you copy my examples into a regression test file or would you like my help with that in some way?

Reply:
Making this change exposes another issue in cast placement.  This branch in `surroundByCast` can silently fail to insert the cast on the current main branch.  On this PR branch before to your change, the insert would succeed, but the rewritten code could be incorrect. After applying your changes, it fails at the error case in `rewriteSourceRange`.
https://github.com/correctcomputation/checkedc-clang/blob/a0988808a3f04802efa64a66579f9fe8ab3078dd/clang/lib/3C/CastPlacement.cpp#L184-L187

Reply:
What's the example?  Is it important to us now (e.g., benchmark tests)?  It would make me feel a little better about spending time trying to fix unusual cases if, in the process, I caught a latent problem in a real-world case. :slightly_smiling_face:  I guess the situation is similar to #423.  If the example didn't actually need the cast in order to pass the type checker, maybe the best thing for now is to skip the cast insertion with a warning.

Reply:
```c
#define NULL ((int*)0)
int foo (int *v);
void bar() {
  foo(NULL);
}
```

Reply:
This bug is the result of two issue in 3C that have been a constant problem. 

1.  Cast insertion makes calls to `getExprConstraintVars` to get the constraint variables for argument expressions. The problem with this is that sometimes `getExprConstraintVars` will need to allocate a new `ConstraintVariable`. If it does this after constraint solving, then the solution for the variables defaults to Checked. This obviously is a problem if the expression would have solved to WILD. To avoid this issue, we store constraint variable sets for expressions in a map indexed by source location. 
2. Expression in macros don't have unique a `PersistentSourceLocation`, so using the source location as a key in a map can result in different expressions being conflated due to source location collision. To avoid this, expression in a macro aren't stored in the map mentioned above. That is what causes the observed issue. Just storing them in the map regardless of colliding source locations doesn't work either; it causes a different bug.

Reply:
Would it fix part 2 if we store the locations of the entire macro call stack in the `PersistentSourceLocation`?  I have an idea of how to do that from the research into the Clang API that I did when I proposed the original change a few days ago, but we probably don't want to block this PR on that fix.  Maybe we should file a separate issue (if we don't already have one)?

Reply:
There should be an issue open for the general problem of non-unique `PersistentSourceLocation`. I don't think anyone has tried storing the whole macro call stack, but it could work. The most important characteristic for the `PersistentSourceLocation` is that it must be consistent for the same expression between the different clang tool runs. In the past I've run into trouble when clang doesn't seem to be entirely deterministic in exactly how it allocates scratch space for macro expansion (#18, #255).

Reply:
Filed #444.

[Back to:](https://github.com/correctcomputation/checkedc-clang/pull/434#discussion_r580443413)
> John, will you copy my examples into a regression test file or would you like my help with that in some way?

Reply:
Tests look good